packages/backend-app-api/src/wiring/BackendInitializer.ts

```diff
@@ -104,53 +102,51 @@ const instanceRegistry = new (class InstanceRegistry {
 function createInstanceMetadataServiceFactory(
   registrations: InternalBackendRegistrations[],
 ) {
-  const installedFeatures = registrations
-    .map(registration => {
-      if (registration.featureType === 'registrations') {
-        return registration
-          .getRegistrations()
-          .map(feature => {
-            if (feature.type === 'plugin') {
-              return Object.defineProperty(
-                {
-                  type: 'plugin',
-                  pluginId: feature.pluginId,
-                },
-                'toString',
-                {
-                  enumerable: false,
-                  configurable: true,
-                  value: () => `plugin{pluginId=${feature.pluginId}}`,
-                },
-              );
-            } else if (feature.type === 'module') {
-              return Object.defineProperty(
-                {
-                  type: 'module',
-                  pluginId: feature.pluginId,
-                  moduleId: feature.moduleId,
-                },
-                'toString',
-                {
-                  enumerable: false,
-                  configurable: true,
-                  value: () =>
-                    `module{moduleId=${feature.moduleId},pluginId=${feature.pluginId}}`,
-                },
-              );
-            }
-            // Ignore unknown feature types.
-            return undefined;
-          })
-          .filter(Boolean) as BackendFeatureMeta[];
+  const installedPlugins: {
+    [pluginId: string]: InstanceMetadataServicePluginInfo;
+  } = {};
+  for (const registration of registrations) {
+    if (registration.featureType === 'registrations') {
+      for (const feature of registration.getRegistrations()) {
+        if (feature.type === 'plugin') {
+          if (!installedPlugins[feature.pluginId]) {
+            installedPlugins[feature.pluginId] = {
+              pluginId: feature.pluginId,
+              modules: [],
+            };
+          }
+        } else if (feature.type === 'module') {
+          if (!installedPlugins[feature.pluginId]) {
+            installedPlugins[feature.pluginId] = {
+              pluginId: feature.pluginId,
+              modules: [],
+            };
+          }
+          installedPlugins[feature.pluginId].modules.push({
+            moduleId: feature.moduleId,
+          });
+        }
       }
-      return [];
-    })
-    .flat();
+    }
+  }
   return createServiceFactory({
-    service: instanceMetadataServiceRef,
-    deps: {},
-    factory: async () => ({ getInstalledFeatures: () => installedFeatures }),
+    service: coreServices.instanceMetadata,
+    deps: {
+      logger: coreServices.rootLogger,
+    },
+    factory: async ({ logger }) => {
+      const instanceMetadata = {
+        getInstalledPlugins: () => Object.values(installedPlugins),
```

Comment:
I'd go as far as constructing the array and then recursively Object.freeze'ing it, but maybe that's paranoid :D